### DISCUSSION TITLE
Refactor TCL-specific string handling & escaping into single function / interface

### DISCUSSION THREAD

**Discussion by umlaeute** [pure-data/pure-data#1694] [Ideas]
This is research/brainstorming for refactoring the communication between the Pd core and the GUI. https://github.com/pure-data/pure-data/pull/1693

_Note: This is **not** a full, open rewrite away from Tcl/Tk but a refactor and abstraction to allow for implementing custom GUIs without worrying about Tcl-specifics. Also, this current work maintains the editing logic within the core._

**Please keep the discussion focused on technical considerations. This is not meant to be a GUI feature wish list!**

---

currently, Core->GUI communication uses `sys_vgui()` to send raw Tcl/Tk commands, where the caller is responsible for building the command with an `snprintf()`-like syntax.

this discussion is about replacing this low-level approach with a somewhat higher-level approach, that delegates the actual formatting to a new (family of) function(s).

for reference, this is what Purr-Data does: https://git.purrdata.net/jwilkes/purr-data#gui-messaging-specification

**Comment by umlaeute** [comment]
short rundown of what Purr Data does:

`sys_vgui()` is replaced with `gui_vmess()`:

```C
void gui_vmess(const char *msg, const char *format, ...);

/* and for building up messages value-by-value */
void gui_start_vmess(const char *msg, const char *format, ...);
void gui_start_array(void);  // start an array
void gui_f(t_float f);       // floating point array element (t_float)
void gui_i(int i);           // integer array element (int)
void gui_s(const char *str); // c string array element
void gui_end_array();        // end an array
void gui_end_vmess();        // terminate the message
```

- `msg` is the actual receiver on the GUI side (a Tcl `proc`)
- `format` is a string that defines the types of the following variadic arguments. the following specifiers are known:
   | id | type | description |
   |---|-------|------------|
   | `f` | `t_float` | floating point value
   | `i` | `int` | integer
   | `s` | `const char*` | c string
   | `x` | `void*` | hexadecimal integer value, with a precision of at least six digits. (hex value is preceded by an 'x', like x123456)

**Reply by umlaeute** [reply]
for object-identifiers (`t_pd*`) we might want to have a separate type-id, e.g. `o`

|id | type | description |
|---|----|---|
|`o`|`t_pd*`| an object identifier

**Reply by umlaeute** [reply]
arrays could be expressed as an (array-id, number-of-elements, values) tuple:

(see https://github.com/pure-data/pure-data/discussions/1694#discussioncomment-2939417 for the id-value syntax)

```C
gui_tvmess("setarray",
   's', "array1",
   'F', 1024, *data,
   0);
```

i wonder, whether this would also need a `stride` (to easily dump `garray_getfloatwords()` to the GUI).
or simply use a special type-id for that case...

|id|type|description|
|---|---|---|
|`F`|`t_float*`|array of floats
|`I`|`int*`|array of ints
|`W`|`t_word*`|array of words

**Reply by Spacechild1** [reply]
> arrays could be expressed as an (array-id, number-of-elements, values) tuple:

I like this! I think it's easier to use (and also more efficient) than PurrData's approach.

**Comment by umlaeute** [comment]
one of the biggest problems i have with `printf()`-formats (and which i see mimicked in `gui_vmess()` and `binbuf_addv()`) is that the format specification is separated from the values, which makes it really hard to read. (and harder to edit): e.g.

```C
      gui_vmess("gui_gobj_draw_io", "xxxiiiiiisiii",
                glist_getcanvas(glist),
                x,
                x,
                onset,
                ypos + x->w.height - 1 + 2 * BORDER,
                onset + IOWIDTH,
                ypos + x->w.height + 2 * BORDER,
                xpos,
                ypos,
                "o",
                i,
                0,
                0);
```

adding a new 5th argument of type `s` involves a lot of counting...

i wonder whether a 0-terminated type-value pair might make more sense (while being equally fast to parse):
```C
      gui_tvmess("gui_gobj_draw_io",
                'x', glist_getcanvas(glist),
                'x', x,
                'x', x,
                'i', onset,
                'i', ypos + x->w.height - 1 + 2 * BORDER,
                'i', onset + IOWIDTH,
                'i', ypos + x->w.height + 2 * BORDER,
                'i', xpos,
                'i', ypos,
                's', "o",
                'i', i,
                'i', 0,
                'i', 0,
                0
);
```

**Reply by Spacechild1** [reply]
> i wonder whether a 0-terminated id-value pair might make more sense (while being equally fast to parse):

I really like this! The only potential "problem" I see is that it is not consistent with the rest of Pd's variadic functions, such as `binbuf_addv` and `pd_vmess`. I guess it's a tradeoff between usability and consistency :-)

**Reply by danomatika** [reply]
I would also agree that I prefer the consistency. How often does one end up changing this code anyway?

**Reply by giuliomoro** [reply]
And when they do change it, is it done/should it be done in a backwards compatible way, only appending new elements to the end? In such case , it would be a non-issue.

**Reply by umlaeute** [reply]
i'm aware of the inconsistency (that's why i mentioned `binbuf_addv`), and I'm not very fond of having inconsistencies either. otoh, i very much prefer a readable interface over a consistent one.

**Comment by umlaeute** [comment]
> How often does one end up changing this code anyway?

> And when they do change it, is it done/should it be done in a backwards compatible way, only appending new elements to the end?

i think that's beside the point.
- the readability problem occurs *every* time somebody is changing the code (and each time is a justification to use a different API)
- appending to the end? not really the problem i'm trying to solve. my original example was mostly to show how it would look like). the motivation was not pure aesthetics but solving real world frustration. e.g. consider https://github.com/pure-data/pure-data/blob/6af3739b2284cfd53ad0ca8eeb2d3c450767e77b/src/g_bang.c#L241-L255 and then try to implement font-selection where you can not only choose from 3 different font (families) but pick any font by name. or due to some multiplication (think zooming) a type has changed from integer to float and suddenly Pd segfaults... happy debugging times (been there, done that, don't want to again)

**Comment by umlaeute** [comment]
and just to expand on why i think the type-value pairs are nice:

- we can use enumerations/constants to make the meaning of each specifier more explicit:
   ```C
   #define GUI_TYPE_FLOAT 'f'
   #define GUI_TYPE_STRING 's'   /* ... */
   gui_tvmess("set_constant",
              GUI_TYPE_STRING, "pi", 
              GUI_TYPE_FLOAT, 3.14,
              GUI_TYPE_END);
   ```

- it allows a simple specification for sending arrays. the purr data approach seems to be suboptimal in this regard.

**Comment by umlaeute** [comment]
i decided to use the `binbuf_addv` compatible approach in #1696 
it turned out, that that since there is no *other* content in the format-specifier (unlike with `sprintf`), it's not *that* ugly to write and read.

**Comment by umlaeute** [comment]
### TODO
somehow handle the `gfxstub_new`, which adds another layer of indirection.

probably easiest would be to replace `gfxstub_new()` with a `gfxstub_newv()` that somehow injects the handle to the graphics responder.
(no idea how to do that yet)
